packages/expo-image-picker/android/src/main/java/expo/modules/imagepicker/ImagePickerModule.java

```diff
@@ -322,440 +154,198 @@ private void launchCameraWithPermissionsGranted(Promise promise, Intent cameraIn
+        packageName,
+        mCameraCaptureURI,
+        Intent.FLAG_GRANT_WRITE_URI_PERMISSION | Intent.FLAG_GRANT_READ_URI_PERMISSION
       );
     }
     mPromise = promise;
     // camera intent needs a content URI but we need a file one
-    cameraIntent.putExtra(MediaStore.EXTRA_OUTPUT, ImagePickerFileUtils.contentUriFromFile(imageFile, getApplication(null)));
-    startActivityOnResult(cameraIntent, REQUEST_LAUNCH_CAMERA, promise);
+    cameraIntent.putExtra(MediaStore.EXTRA_OUTPUT, ImagePickerFileUtils.contentUriFromFile(imageFile, getApplication()));
+    startActivityOnResult(cameraIntent, ImagePickerConstance.REQUEST_LAUNCH_CAMERA, promise);
   }
 
   // NOTE: Currently not reentrant / doesn't support concurrent requests
   @ExpoMethod
   public void launchImageLibraryAsync(final Map<String, Object> options, final Promise promise) {
-    if (!readOptions(options, promise)) {
+    mPickerOptions = PickerOptions.fromMap(options, promise);
+    if (mPickerOptions == null) {
       return;
     }
 
     Intent libraryIntent = new Intent();
 
-    if (mediaTypes.equals("Images")) {
-      libraryIntent.setType("image/*");
-    } else if (mediaTypes.equals("Videos")) {
-      libraryIntent.setType("video/*");
-    } else if (mediaTypes.equals("All")) {
-      libraryIntent.setType("*/*");
-      String[] mimetypes = {"image/*", "video/*"};
-      libraryIntent.putExtra(Intent.EXTRA_MIME_TYPES, mimetypes);
+    switch (mPickerOptions.getMediaTypes()) {
+      case "Images":
+        libraryIntent.setType("image/*");
+        break;
+      case "Videos":
+        libraryIntent.setType("video/*");
+        break;
+      case "All":
+        libraryIntent.setType("*/*");
+        libraryIntent.putExtra(Intent.EXTRA_MIME_TYPES, new String[]{"image/*", "video/*"});
+        break;
     }
 
     mPromise = promise;
 
     libraryIntent.setAction(Intent.ACTION_GET_CONTENT);
-    startActivityOnResult(libraryIntent, REQUEST_LAUNCH_IMAGE_LIBRARY, promise);
+    startActivityOnResult(libraryIntent, ImagePickerConstance.REQUEST_LAUNCH_IMAGE_LIBRARY, promise);
   }
 
-  public void onActivityResult(final int requestCode, final int resultCode, final Intent intent) {
-    if (requestCode == CropImage.CROP_IMAGE_ACTIVITY_REQUEST_CODE) {
-      if (mLaunchedCropper) {
-        mLaunchedCropper = false;
-        final Promise promise = mPromise;
-        mPromise = null;
-        Bundle exifData = mExifData;
-        mExifData = null;
-
-        if (promise == null) {
-          return;
-        }
-        if (resultCode != Activity.RESULT_OK) {
-          Bundle response = new Bundle();
-          response.putBoolean("cancelled", true);
-          promise.resolve(response);
-          return;
-        }
-
-        handleCropperResult(intent, promise, exifData);
-      }
-      return;
+  public void startCropIntent(Uri uri, String type, int requestCode) throws IOException {
+    String extension = ".jpg";
```

Comment:
Is type expected to be of format `image/(png|gif|jpg)`?

Could we verify it is of given format? And check against endsWith? I am afraid those checks are vulnerable to future errors.

Reply:
No, the format could be different. However, then we will try to convert it to a `jpg`. So, we can't check it using `endsWith`.